include/crocoddyl/core/integrator/rk4.hpp

```diff
@@ -1,14 +1,156 @@
+
+  const boost::shared_ptr<DifferentialActionModelAbstract>& get_differential() const;
+  const Scalar& get_dt() const;
+
+  void set_dt(const Scalar& dt);
+  void set_differential(boost::shared_ptr<DifferentialActionModelAbstract> model);
+
+ protected:
+  using Base::has_control_limits_;  //!< Indicates whether any of the control limits are active
+  using Base::nr_;                  //!< Dimension of the cost residual
+  using Base::nu_;                  //!< Control dimension
+  using Base::state_;               //!< Model of the state
+  using Base::u_lb_;                //!< Lower control limits
+  using Base::u_ub_;                //!< Upper control limits
+  using Base::unone_;               //!< Neutral state
+
+ private:
+  boost::shared_ptr<DifferentialActionModelAbstract> differential_;
+  Scalar time_step_;
+  std::vector<Scalar> rk4_c_;
+  bool with_cost_residual_;
+  bool enable_integration_;
+};
+
+template <typename _Scalar>
+struct IntegratedActionDataRK4Tpl : public ActionDataAbstractTpl<_Scalar> {
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef ActionDataAbstractTpl<Scalar> Base;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  template <template <typename Scalar> class Model>
+  explicit IntegratedActionDataRK4Tpl(Model<Scalar>* const model) : Base(model) {
+    const std::size_t& ndx = model->get_state()->get_ndx();
+    const std::size_t& nx = model->get_state()->get_nx();
+
+    const std::size_t& nu = model->get_nu();
+    for (std::size_t i = 0; i < 4; ++i) {
+      differential.push_back(
+          boost::shared_ptr<DifferentialActionDataAbstractTpl<Scalar>>(model->get_differential()->createData()));
+    }
+
+    dx = VectorXs::Zero(ndx);
+    integral = std::vector<Scalar>(4, Scalar(0.));
+
+    ki = std::vector<VectorXs>(4, VectorXs::Zero(ndx));
+    y = std::vector<VectorXs>(4, VectorXs::Zero(nx));
+    dx_rk4 = std::vector<VectorXs>(4, VectorXs::Zero(ndx));
+
+    dki_dx = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, ndx));
+    dki_du = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, nu));
+    dyi_dx = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, ndx));
+    dyi_du = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, nu));
+    dki_dy = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, ndx));
+
+    dli_dx = std::vector<VectorXs>(4, VectorXs::Zero(ndx));
+    dli_du = std::vector<VectorXs>(4, VectorXs::Zero(nu));
+    ddli_ddx = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, ndx));
+    ddli_ddu = std::vector<MatrixXs>(4, MatrixXs::Zero(nu, nu));
+    ddli_dxdu = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, nu));
+    Luu_partialx = std::vector<MatrixXs>(4, MatrixXs::Zero(nu, nu));
+    Lxx_partialx = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, ndx));
+    Lxx_partialu = std::vector<MatrixXs>(4, MatrixXs::Zero(ndx, nu));
+  }
+  ~IntegratedActionDataRK4Tpl() {}
```

Comment:
Define the destructor as virtual
```cpp
virtual ~IntegratedActionDataRK4Tpl() {}
```

Reply:
This is the data. In case this has to be virtual, should I change also the `IntegratedActionDataEulerTpl` ?

Reply:
Yes, please do it for the Euler integrator as well.

Reply:
done. Which is the reason for doing the destructor virtual?

Reply:
So that if there are any classes that inherit from this class, and pointers are resolved dynamically, the destructor of the derived class is used and not of the base class.

Reply:
In top of what Rohan said, if you don't call the derived destructor it will occurs memory leak. You might want to expect 5 minutes reading the following link: https://www.quantstart.com/articles/C-Virtual-Destructors-How-to-Avoid-Memory-Leaks/